File: Binary Search/Count_occurences.java
public class Count_occurences{
    public static void main(String [] args){
        int [] arr = {1, 1, 2, 2, 2, 2, 3};
        int x = 2;

        int first = LowerBound(arr, x);
        int last = UpperBound(arr, x) - 1;

        if (first == arr.length || arr[first] != x) {
            System.out.println("0");
        }
        else{
            System.out.println(last - first + 1);
        }
    }

    public static int LowerBound(int [] arr, int x){
        int low = 0, high = arr.length - 1;
        int ans = arr.length;

        while (low <= high) { 
            int mid = (low + high)/2;

            if (arr[mid] >= x) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }

    public static int UpperBound(int [] arr, int x){
        int low = 0, high = arr.length - 1;
        int ans = arr.length;

        while (low <= high) {
            int mid = (low + high)/2;

            if (arr[mid] > x) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
}

File: Binary Search/BS on 1D Arrays/Python/Count_Occurences.py
def count_occurences(nums, x):
    first, last = first_last_occurence(nums, x)

    if first == -1:
        return 0
    else:
        return last - first + 1


def first_last_occurence(nums, x):
    first = LowerBound(nums, x)
    last = UpperBound(nums, x) - 1

    if first == -1:
        return [-1, -1]
    else:
        return [first, last]

    
def LowerBound(nums, x):
    n = len(nums)

    low = 0
    high = n - 1

    ans = n

    while low <= high:
        mid = int((low + high)/2)

        if nums[mid] >= x:
            ans = mid
            high = mid - 1
        else:
            low = mid + 1

    return ans


def UpperBound(nums, x):
    n = len(nums)

    low = 0
    high = n - 1

    ans = n

    while low <= high:
        mid = int((low + high)/2)

        if nums[mid] > x:
            ans = mid
            high = mid - 1
        else:
            low = mid + 1

    return ans



nums = [2, 4, 6, 8, 8, 8, 11, 13]
x = 8
print(count_occurences(nums, x))